### PR TITLE
Update upsert batch version

### PR DESCRIPTION
**Pull request by jshencode**
add version number to upsert batch
and update upsert batch format to include arrival time

### CODE REVIEW

File: memstore/common/upsert_batch_builder.go [hunk 1]
```diff
@@ -387,8 +391,63 @@ func (u *UpsertBatchBuilder) SetValue(row int, col int, value interface{}) error
 	return u.columns[col].SetValue(row, value)
 }
 
-// ToByteArray produces a serialized UpsertBatch in byte array.
-func (u UpsertBatchBuilder) ToByteArray() ([]byte, error) {
+// ToByteArrayNew will be used for new upsert batch version
+func (u UpsertBatchBuilder) ToByteArrayNew() ([]byte, error) {
+	// Create buffer.
+	numCols := len(u.columns)
+	size := 4 // version number
```
Comment (voyager-dw) [inline memstore/common/upsert_batch_builder.go]: maybe: size for magic version number

File: memstore/common/upsert_batch_builder.go [hunk 2]
```diff
@@ -387,8 +391,63 @@ func (u *UpsertBatchBuilder) SetValue(row int, col int, value interface{}) error
...
+	headerSize := 12 + ColumnHeaderSize(numCols)
```
Comment (voyager-dw) [inline memstore/common/upsert_batch_builder.go]: add comment for 12?

File: memstore/common/upsert_batch_builder.go [hunk 3]
```diff
@@ -387,8 +391,63 @@ func (u *UpsertBatchBuilder) SetValue(row int, col int, value interface{}) error
...
+		}
+	}
+
+	return buffer, nil
+}
+
+func (u UpsertBatchBuilder) toByteArrayOld() ([]byte, error)  {
 	// Create buffer.
 	numCols := len(u.columns)
 	headerSize := 8 + ColumnHeaderSize(numCols)
```
Comment (voyager-dw) [inline memstore/common/upsert_batch_builder.go]: add comment for 8?

File: memstore/common/upsert_batch_builder.go [hunk 4]
```diff
@@ -387,8 +391,63 @@ func (u *UpsertBatchBuilder) SetValue(row int, col int, value interface{}) error
...
+	}
+	if err := writer.WriteUint32(uint32(utils.Now().Unix()), 12); err != nil  {
+		return nil, utils.StackError(err, "Failed to write arrival time")
+	}
+	writer.SkipBytes(headerSize + 4)
+
+	header := NewUpsertBatchHeader(buffer[16:headerSize], numCols)
+	// Write per column data their headers.
+	for i, column := range u.columns {
+		if err := header.WriteColumnID(column.columnID, i); err != nil {
```
Comment (voyager-dw) [inline memstore/common/upsert_batch_builder.go]: shall we push the error message down to each function?

File: memstore/common/upsert_batch_builder.go [hunk 5]
```diff
@@ -387,8 +391,63 @@ func (u *UpsertBatchBuilder) SetValue(row int, col int, value interface{}) error
...
+		}
+		if err := header.WriteColumnOffset(writer.GetOffset(), i+1); err != nil {
+			return nil, utils.StackError(err, "Failed to write end offset for column %d", i)
+		}
+	}
+
+	return buffer, nil
+}
+
+func (u UpsertBatchBuilder) toByteArrayOld() ([]byte, error)  {
```
Comment (voyager-dw) [inline memstore/common/upsert_batch_builder.go]: do we still need this func?

File: memstore/common/upsert_batch_builder.go [hunk 6]
```diff
@@ -387,8 +391,63 @@ func (u *UpsertBatchBuilder) SetValue(row int, col int, value interface{}) error
 	return u.columns[col].SetValue(row, value)
 }
 
-// ToByteArray produces a serialized UpsertBatch in byte array.
-func (u UpsertBatchBuilder) ToByteArray() ([]byte, error) {
+// ToByteArrayNew will be used for new upsert batch version
+func (u UpsertBatchBuilder) ToByteArrayNew() ([]byte, error) {
```
Comment (lucafuji) [inline memstore/common/upsert_batch_builder.go]: can we have this function call old ToByteArray?

File: memstore/common/upsert_batch_builder.go (hunk 5 above)
Comment (jshencode) [reply memstore/common/upsert_batch_builder.go]: yes. both func needed before migration is done.
the server side need to know how to handle both version

**Pull request merged**